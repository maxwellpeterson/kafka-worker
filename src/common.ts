export interface Env {
  // The hostname of this worker
  HOSTNAME: string;
  CLUSTER: DurableObjectNamespace;
  // Example binding to KV. Learn more at https://developers.cloudflare.com/workers/runtime-apis/kv/
  // MY_KV_NAMESPACE: KVNamespace;
  //
  // Example binding to R2. Learn more at https://developers.cloudflare.com/workers/runtime-apis/r2/
  // MY_BUCKET: R2Bucket;
}

export type ValueOf<T> = T[keyof T];
