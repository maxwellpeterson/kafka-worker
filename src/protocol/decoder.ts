import {
  Int16,
  Int32,
  int16Size,
  int32Size,
  String,
  KafkaArray,
  ErrorCode,
  validErrorCode,
  Int64,
  int64Size,
} from "src/protocol/common";

export class Decoder {
  private readonly view: DataView;
  private offset: number;

  constructor(buffer: ArrayBuffer) {
    this.view = new DataView(buffer);
    this.offset = 0;
  }

  readInt16(): Int16 {
    const value = this.view.getInt16(this.offset);
    this.offset += int16Size;
    return value;
  }

  readEnum<T extends Int16>(predicate: (value: Int16) => value is T): T {
    const value = this.readInt16();
    if (!predicate(value)) {
      throw new Error(`Invalid enum value: ${value}`);
    }
    return value;
  }

  readInt32(): Int32 {
    const value = this.view.getInt32(this.offset);
    this.offset += int32Size;
    return value;
  }

  readInt64(): Int64 {
    const value = this.view.getBigInt64(this.offset);
    this.offset += int64Size;
    return value;
  }

  readString(): String {
    const size = this.readInt16();
    if (size === -1) {
      // I don't think this should ever happen...
      throw new Error("Unexpected null string!");
    }
    const stringBuffer = this.view.buffer.slice(
      this.offset,
      this.offset + size
    );
    const value = new TextDecoder().decode(stringBuffer);
    this.offset += size;
    return value;
  }

  readKafkaArray<T>(readElement: (index: number) => T): KafkaArray<T> {
    const size = this.readInt32();
    if (size === -1) {
      return null;
    }
    const values = new Array<T>(size);
    for (let i = 0; i < size; i++) {
      values[i] = readElement(i);
    }
    return values;
  }

  // In several apis, there is no semantic difference between empty and null
  // arrays, and converting null arrays to empty arrays eliminates unnecessary
  // code pollution from null checking
  readArray<T>(readElement: (index: number) => T): Array<T> {
    return this.readKafkaArray(readElement) ?? [];
  }

  readMessageSet(size: Int32): ArrayBuffer {
    const buffer = this.view.buffer.slice(this.offset, this.offset + size);
    this.offset += size;
    return buffer;
  }
}

export class KafkaRequestDecoder extends Decoder {
  constructor(buffer: ArrayBuffer) {
    super(buffer);

    const expectedSize = this.readInt32();
    const actualSize = buffer.byteLength - int32Size;

    if (expectedSize !== actualSize) {
      throw new Error(
        `Message length does not match size header: expected ${expectedSize} but got ${actualSize}`
      );
    }
  }
}
