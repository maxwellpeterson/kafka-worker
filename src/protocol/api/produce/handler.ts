import { Env } from "src/common";
import {
  ProduceRequest,
  ProduceResponse,
} from "src/protocol/api/produce/types";

export const handleProduceRequest = async (
  env: Env,
  request: ProduceRequest
): Promise<ProduceResponse | null> => {
  return null;
};
