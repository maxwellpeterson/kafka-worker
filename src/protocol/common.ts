import { ValueOf } from "src/common";

// These don't add any type safety, but help with labeling
export type Int16 = number;
export type Int32 = number;
export type Int64 = bigint;

export const int16Size = 2;
export const int32Size = 4;
export const int64Size = 8;

export type KafkaArray<T> = T[] | null;

export interface RequestMetadata {
  apiVersion: Int16;
  correlationId: Int32;
  clientId: string;
}
export const generateEnumPredicate =
  <E extends Record<string, V>, V>(enumObj: E) =>
  (value: V): value is ValueOf<E> =>
    Object.values<V>(enumObj).includes(value);

// ApiKey is an Int16
export const ApiKey = {
  Produce: 0,
  // Fetch: 1,
  // ListOffsets: 2,
  Metadata: 3,
} as const;
export type ApiKey = ValueOf<typeof ApiKey>;
export const validApiKey = generateEnumPredicate(ApiKey);

// ErrorCode is an Int16
export const ErrorCode = {
  None: 0,
  UnknownTopicOrPartition: 3,
} as const;
export type ErrorCode = ValueOf<typeof ErrorCode>;
export const validErrorCode = generateEnumPredicate(ErrorCode);

// Acks is an Int16
export const Acks = {
  None: 0,
  Leader: 1,
  FullISR: -1,
} as const;
export type Acks = ValueOf<typeof Acks>;
export const validAcks = generateEnumPredicate(Acks);
