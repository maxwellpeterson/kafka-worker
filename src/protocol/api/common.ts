import { Env } from "src/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

export type EntrypointFn<Output> = (
  env: Env,
  version: number,
  decoder: Decoder,
  encoder: Encoder
) => Promise<Output>;

export type CheckVersionFn = (version: number) => void;
export type DecodeFn<Req> = (decoder: Decoder) => Req;
export type HandleFn<Req, Res> = (env: Env, request: Req) => Promise<Res>;
export type EncodeFn<Res, Output> = (encoder: Encoder, response: Res) => Output;

export const generateEntrypointFn =
  <Req, Res, Output>(
    apiName: string,
    checkVersion: CheckVersionFn,
    decode: DecodeFn<Req>,
    handle: HandleFn<Req, Res>,
    encode: EncodeFn<Res, Output>
  ): EntrypointFn<Output> =>
  async (
    env: Env,
    version: number,
    decoder: Decoder,
    encoder: Encoder
  ): Promise<Output> => {
    checkVersion(version);

    const request = decode(decoder);
    console.log(
      `Received ${apiName} request: ${JSON.stringify(request, null, 2)}`
    );

    const response = await handle(env, request);
    console.log(
      `Sending ${apiName} response: ${JSON.stringify(response, null, 2)}`
    );

    return encode(encoder, response);
  };

export const generateCheckVersionFn =
  (apiName: string, expected: number): CheckVersionFn =>
  (version: number) => {
    if (version !== expected) {
      throw new Error(
        `Unsupported version of ${apiName} api: expected ${expected} but got ${version}`
      );
    }
  };
