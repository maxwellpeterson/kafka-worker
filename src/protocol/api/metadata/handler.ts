import { Env } from "src/common";
import {
  MetadataRequest,
  MetadataResponse,
} from "src/protocol/api/metadata/types";
import { KafkaArray, KafkaString } from "src/protocol/common";
import { fetchClusterMetadata } from "src/state/cluster";

export const handleMetadataRequest = async (
  env: Env,
  request: MetadataRequest
): Promise<MetadataResponse> => {
  return await fetchClusterMetadata(env, cleanTopics(request.topics));
};

const cleanTopics = (topics: KafkaArray<KafkaString>): string[] => {
  if (topics === null) {
    return [];
  }
  return topics.filter((topic): topic is string => topic !== null);
};
