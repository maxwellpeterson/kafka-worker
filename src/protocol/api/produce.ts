import { Acks, ErrorCode, Int32, Int64 } from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

// Produce Request (Version: 0) => acks timeout_ms [topic_data]
//   acks => INT16
//   timeout_ms => INT32
//   topic_data => name [partition_data]
//     name => STRING
//     partition_data => index records
//       index => INT32
//       records => RECORDS
//
// https://kafka.apache.org/protocol.html#The_Messages_Produce
//
// There some variation here between the current protocol spec and
// the older protocol guide:
//
// https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol
//
// Here we follow the older description of message set (now called record batch).

export interface ProduceRequest {
  acks: Acks;
  timeoutMs: Int32;
  topics: TopicData[];
}

export const decodeProduceRequest = (decoder: Decoder): ProduceRequest => {
  return {
    acks: decoder.readAcks(),
    timeoutMs: decoder.readInt32(),
    topics: decoder.readArray(decodeTopicData(decoder)),
  };
};

export interface TopicData {
  name: string;
  partitions: PartitionData[];
}

const decodeTopicData = (decoder: Decoder) => (): TopicData => {
  return {
    name: decoder.readString(),
    partitions: decoder.readArray(decodePartitionData(decoder)),
  };
};

export interface PartitionData {
  index: Int32;
  messageSetSize: Int32;
  messageSet: ArrayBuffer;
}

const decodePartitionData = (decoder: Decoder) => (): PartitionData => {
  const data = {
    index: decoder.readInt32(),
    messageSetSize: decoder.readInt32(),
  };
  return { ...data, messageSet: decoder.readBuffer(data.messageSetSize) };
};

// Produce Response (Version: 0) => [responses]
//   responses => name [partition_responses]
//   name => STRING
//   partition_responses => index error_code base_offset
//     index => INT32
//     error_code => INT16
//     base_offset => INT64
//
// https://kafka.apache.org/protocol.html#The_Messages_Produce

export interface ProduceResponse {
  topics: TopicResponse[];
}

export const encodeProduceResponse = (
  encoder: Encoder,
  response: ProduceResponse
): ArrayBuffer => {
  return encoder
    .writeArray(response.topics, encodeTopicResponse(encoder))
    .buffer();
};

export interface TopicResponse {
  name: string;
  partitions: PartitionResponse[];
}

const encodeTopicResponse =
  (encoder: Encoder) =>
  (topic: TopicResponse): Encoder => {
    return encoder
      .writeString(topic.name)
      .writeArray(topic.partitions, encodePartitionResponse(encoder));
  };

export interface PartitionResponse {
  index: Int32;
  errorCode: ErrorCode;
  baseOffset: Int64;
}

const encodePartitionResponse =
  (encoder: Encoder) =>
  (partition: PartitionResponse): Encoder => {
    return encoder
      .writeInt32(partition.index)
      .writeEnum(partition.errorCode)
      .writeInt64(partition.baseOffset);
  };
