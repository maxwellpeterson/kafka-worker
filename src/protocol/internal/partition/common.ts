import { ValueOf } from "src/common";
import {
  ApiKey,
  Int16,
  Int32,
  generateEnumPredicate,
  int32Size,
} from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

// Partition Request (Version: 0) => header request_message
//   header => api_key api_version correlation_id client_id
//     api_key => INT16
//     api_version => INT16
//     correlation_id => INT32
//     client_id => STRING
//   request_message => PARTITION_PRODUCE_REQUEST

export interface PartitionRequestHeader {
  apiKey: PartitionApiKey;
  apiVersion: Int16;
  correlationId: Int32;
  clientId: string;
}

export const encodePartitionRequestHeader = (
  encoder: Encoder,
  header: PartitionRequestHeader
) => {
  encoder.writeEnum(header.apiKey);
  encoder.writeInt16(header.apiVersion);
  encoder.writeInt32(header.correlationId);
  encoder.writeString(header.clientId);
};

export const decodePartitionRequestHeader = (
  decoder: Decoder
): PartitionRequestHeader => {
  return {
    apiKey: decoder.readEnum(validPartitionApiKey),
    apiVersion: decoder.readInt16(),
    correlationId: decoder.readInt32(),
    clientId: decoder.readString(),
  };
};

// PartitionApiKey is an Int16
export const PartitionApiKey = {
  Produce: ApiKey.Produce,
} as const;
export type PartitionApiKey = ValueOf<typeof PartitionApiKey>;
const validPartitionApiKey = generateEnumPredicate(PartitionApiKey);

// Partition Response (Version: 0) => correlation_id response_message
//   correlation_id => INT32
//   response_message => PARTITION_PRODUCE_RESPONSE

export class PartitionResponseEncoder extends Encoder {
  constructor(correlationId: Int32, initialBufferSize = 64) {
    super(initialBufferSize + int32Size);
    this.writeInt32(correlationId);
  }
}
