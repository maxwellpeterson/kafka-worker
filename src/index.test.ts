import handler from "src/index";
import { ApiKey, Int32 } from "src/protocol/common";
import { KafkaDecoder, KafkaRequestEncoder } from "src/protocol/kafka/common";
import {
  KafkaMetadataResponse,
  decodeKafkaMetadataResponse,
  encodeKafkaMetadataRequest,
} from "src/protocol/kafka/metadata";
import { KafkaProduceResponse } from "src/protocol/kafka/produce";
import { base64 } from "src/protocol/test-utils";
import { Decoder } from "./protocol/decoder";

class GatewayConn {
  private readonly socket: WebSocket;
  private readonly unhandled: (ArrayBuffer | string)[];
  private handleMessage?: (message: ArrayBuffer) => void;

  constructor() {
    const env = getMiniflareBindings();

    const response = handler.fetch(
      new Request("https://fetch.handler", {
        headers: { Upgrade: "websocket" },
      }),
      env
    );
    if (!response.ok) {
      throw new Error(`Initial request failed with status ${response.status}`);
    }

    const socket = response.webSocket;
    if (socket === null) {
      throw new Error("Initial response does not include WebSocket");
    }

    socket.accept();
    this.socket = socket;
    this.unhandled = [];

    this.socket.addEventListener("message", (event) => {
      if (typeof event.data === "string" || this.handleMessage === undefined) {
        this.unhandled.push(event.data);
        return;
      }
      this.handleMessage(event.data);
    });
  }

  request(buffer: ArrayBuffer): Promise<ArrayBuffer> {
    if (this.unhandled.length > 0) {
      throw new Error(
        `Can't send next message when there are ${this.unhandled.length} unhandled messages`
      );
    }

    if (this.handleMessage !== undefined) {
      throw new Error(
        "Can't send next message when there is a request in flight"
      );
    }

    this.socket.send(buffer);

    return new Promise((resolve) => {
      this.handleMessage = (message: ArrayBuffer) => {
        this.handleMessage = undefined;
        resolve(message);
      };
    });
  }

  close() {
    if (this.unhandled.length > 0) {
      throw new Error(
        `Can't close connection when there are ${this.unhandled.length} unhandled messages`
      );
    }

    this.socket.close();
  }
}

type DecodeFunc<T> = (decoder: Decoder) => T;
type RequestResponse<T> = [Int32, ArrayBuffer, DecodeFunc<T>];
type TestCase<T> = [string, RequestResponse<T>[]];

const makeMetadataPair = (
  correlationId: Int32,
  topics: string[]
): RequestResponse<KafkaMetadataResponse> => {
  const encoder = new KafkaRequestEncoder({
    apiKey: ApiKey.Metadata,
    apiVersion: 0,
    correlationId,
    clientId: null,
  });
  const request = encodeKafkaMetadataRequest(encoder, {
    topics,
  });
  return [correlationId, request, decodeKafkaMetadataResponse];
};

describe("Metadata API", () => {
  const cases: TestCase<KafkaMetadataResponse | KafkaProduceResponse>[] = [
    ["fetch all topics", [makeMetadataPair(5, [])]],
    ["fetch specific topic", [makeMetadataPair(5, ["test-topic"])]],
    ["fetch nonexistent topic", [makeMetadataPair(5, ["other-topic"])]],
    [
      "fetch specific topic and nonexistent topic",
      [makeMetadataPair(5, ["test-topic", "other-topic"])],
    ],
  ];

  test.each(cases)("%s", async (_name, pairs) => {
    const gateway = new GatewayConn();

    for (const [correlationId, request, decodeResponse] of pairs) {
      const response = await gateway.request(request);
      expect(base64(response)).toMatchSnapshot();

      const decoder = new KafkaDecoder(response);
      expect(decoder.readInt32()).toEqual(correlationId);

      const decoded = decodeResponse(decoder);
      expect(decoded).toMatchSnapshot();
    }

    gateway.close();
  });
});
