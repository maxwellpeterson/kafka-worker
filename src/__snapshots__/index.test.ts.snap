// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metadata API fetch all topics 1`] = `"AAAAQwAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAEAAAAKdGVzdC10b3BpYwAAAAEAAAAAA+cAAAFNAAAAAAAAAAA="`;

exports[`Metadata API fetch all topics 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 0,
      "name": "test-topic",
      "partitions": Array [
        Object {
          "errorCode": 0,
          "isrNodes": Array [],
          "leaderId": 333,
          "partitionIndex": 999,
          "replicaNodes": Array [],
        },
      ],
    },
  ],
}
`;

exports[`Metadata API fetch nonexistent topic 1`] = `"AAAAMgAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAEAAwALb3RoZXItdG9waWMAAAAA"`;

exports[`Metadata API fetch nonexistent topic 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 3,
      "name": "other-topic",
      "partitions": Array [],
    },
  ],
}
`;

exports[`Metadata API fetch specific topic 1`] = `"AAAAQwAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAEAAAAKdGVzdC10b3BpYwAAAAEAAAAAA+cAAAFNAAAAAAAAAAA="`;

exports[`Metadata API fetch specific topic 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 0,
      "name": "test-topic",
      "partitions": Array [
        Object {
          "errorCode": 0,
          "isrNodes": Array [],
          "leaderId": 333,
          "partitionIndex": 999,
          "replicaNodes": Array [],
        },
      ],
    },
  ],
}
`;

exports[`Metadata API fetch specific topic and nonexistent topic 1`] = `"AAAAVgAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAIAAAAKdGVzdC10b3BpYwAAAAEAAAAAA+cAAAFNAAAAAAAAAAAAAwALb3RoZXItdG9waWMAAAAA"`;

exports[`Metadata API fetch specific topic and nonexistent topic 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 0,
      "name": "test-topic",
      "partitions": Array [
        Object {
          "errorCode": 0,
          "isrNodes": Array [],
          "leaderId": 333,
          "partitionIndex": 999,
          "replicaNodes": Array [],
        },
      ],
    },
    Object {
      "errorCode": 3,
      "name": "other-topic",
      "partitions": Array [],
    },
  ],
}
`;
